app/src/main/java/com/duckduckgo/app/browser/applinks/AppLinksHandler.kt

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2021 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.browser.applinks
+
+import android.os.Build
+import javax.inject.Inject
+
+interface AppLinksHandler {
+    fun handleAppLink(isRedirect: Boolean, isForMainFrame: Boolean, launchAppLink: () -> Unit): Boolean
+    fun handleNonHttpAppLink(isRedirect: Boolean, launchNonHttpAppLink: () -> Unit): Boolean
+    fun enterBrowserState()
+    fun reset()
+}
+
+class DuckDuckGoAppLinksHandler @Inject constructor() : AppLinksHandler {
+
+    var appLinkOpenedInBrowser = false
+
+    override fun handleAppLink(isRedirect: Boolean, isForMainFrame: Boolean, launchAppLink: () -> Unit): Boolean {
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N || isRedirect && appLinkOpenedInBrowser || !isForMainFrame) {
+            return false
+        }
+        launchAppLink()
+        return true
+    }
+
+    override fun handleNonHttpAppLink(isRedirect: Boolean, launchNonHttpAppLink: () -> Unit): Boolean {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && isRedirect && appLinkOpenedInBrowser) {
+            return true
```

Comment:
For non-HTTP links, like market:// or intent://, `true` is returned so that they are ignored. If `false` is returned for those links, they will open in the browser and the user will see the `WebView` error page because they can't be handled. `false` is returned for HTTP app links since the `WebView` can open those.